src/main/java/redis/clients/jedis/JedisClusterInfoCache.java

```diff
@@ -36,22 +45,80 @@ public class JedisClusterInfoCache {
+
+  class TopologyRefreshTask implements Runnable {
+    @Override
+    public void run() {
+      logger.debug("Cluster topology refresh run, old nodes: {}", nodes.keySet());
+      renewClusterSlots(null);
+      logger.debug("Cluster topology refresh run, new nodes: {}", nodes.keySet());
+    }
+  }
+
   public JedisClusterInfoCache(final JedisClientConfig clientConfig, final Set<HostAndPort> startNodes) {
-    this(clientConfig, null, startNodes);
+    this(clientConfig, null, startNodes, DEFAULT_TOPOLOGY_REFRESH_ENABLED, DEFAULT_TOPOLOGY_REFRESH_PERIOD);
   }
 
   public JedisClusterInfoCache(final JedisClientConfig clientConfig,
       final GenericObjectPoolConfig<Connection> poolConfig, final Set<HostAndPort> startNodes) {
+    this(clientConfig, poolConfig, startNodes, DEFAULT_TOPOLOGY_REFRESH_ENABLED, DEFAULT_TOPOLOGY_REFRESH_PERIOD);
+  }
+
+  public JedisClusterInfoCache(final JedisClientConfig clientConfig,
+      final GenericObjectPoolConfig<Connection> poolConfig, final Set<HostAndPort> startNodes,
+      final boolean topologyRefreshEnabled, final Duration topologyRefreshPeriod) {
     this.poolConfig = poolConfig;
     this.clientConfig = clientConfig;
     this.startNodes = startNodes;
+    this.topologyRefreshEnabled = topologyRefreshEnabled;
+    this.topologyRefreshPeriod = topologyRefreshPeriod;
+    if (topologyRefreshEnabled) {
+      logger.info("Cluster topology refresh start, period: {}, startNodes: {}",
+          topologyRefreshPeriod.toString(), startNodes);
+      topologyRefreshExecutor = Executors.newSingleThreadScheduledExecutor();
+      topologyRefreshExecutor.scheduleWithFixedDelay(new TopologyRefreshTask(), topologyRefreshPeriod.toMillis(),
+          topologyRefreshPeriod.toMillis(), TimeUnit.MILLISECONDS);
+    }
+  }
+
+  /**
+   * Check whether the number and order of slots in the cluster topology are equal to CLUSTER_HASHSLOTS
+   * @param slotsInfo the cluster topology
+   * @return if slots is ok, return true, elese return false.
+   */
+  private boolean checkClusterSlotSequence(List<Object> slotsInfo) {
+    List<Integer> slots = new ArrayList<>();
+    for (Object slotInfoObj : slotsInfo) {
+      List<Object> slotInfo = (List<Object>)slotInfoObj;
+      slots.addAll(getAssignedSlotArray(slotInfo));
+    }
+    Collections.sort(slots);
+    if (slots.size() != Protocol.CLUSTER_HASHSLOTS) {
+      return false;
+    }
+    for (int i = 0; i < Protocol.CLUSTER_HASHSLOTS; ++i) {
+      if (i != slots.get(i)) {
+        return false;
+      }
+    }
+    return true;
   }
 
   public void discoverClusterNodesAndSlots(Connection jedis) {
     List<Object> slotsInfo = executeClusterSlots(jedis);
     if (slotsInfo.isEmpty()) {
       throw new JedisClusterOperationException("Cluster slots list is empty.");
     }
+    if (!checkClusterSlotSequence(slotsInfo)) {
+      throw new JedisClusterOperationException("Cluster slots have holes.");
+    }
```

Comment:
IIRC, we accept that cluster slots may have holes... (at least during initialization).

Reply:
Can you describe specifically the circumstances under which holes are allowed?

Reply:
People want to start their Java application and running without Redis cluster is ready!!

Yes, people do/want weird things.

Reply:
What if we use [`jedis.cluster.initNoError`](https://github.com/redis/jedis/blob/5c6b87bdec4c6171cdcad358b5aa29c5d423f297/src/main/java/redis/clients/jedis/providers/ClusterConnectionProvider.java#L22) here as well?

Perhaps move it to a generally accessible place (JedisCluster?) in a generally accessible way (a variable?)?

Reply:
I moved `INIT_NO_ERROR_PROPERTY` to `JedisCluster` and added `INIT_NO_ERROR_PROPERTY` judgment in `JedisClusterInfoCache`.